lib/spack/spack/version.py

```diff
@@ -627,7 +627,20 @@ def satisfies(self, other):
         self_cmp = self._cmp(other.ref_lookup)
         other_cmp = other._cmp(self.ref_lookup)
 
-        if other.is_ref:
+        if self.is_ref and other.is_ref:
+            if self.ref != other.ref:
+                return False
+            elif self.user_supplied_reference and other.user_supplied_reference:
+                return self.ref_version == other.ref_version
+            elif other.user_supplied_reference:
+                return False
+            else:
+                # In this case, 'other' does not supply a version equivalence
+                # with "=" and the commit strings are equal. 'self' may specify
+                # a version equivalence, but that is extra info and will
+                # satisfy no matter what it is.
+                return True
```

Comment:
This is very bug prone. If no version is imposed by the user, it's looked up and something is assigned. So you can have `v1 = abcdef=1.2.3` and `v2 = abcdef` for which `v1.satisfies(v2)` but `v2 < v1` (as it causes a lookup).

I maintain that we should just remove this lookup nonsense, it's just introducing too many edge cases and is generally useless.

Reply:
I'd rather have a centralized conversation on this (if there is such a conversation, can you link to it here?).

In this particular branch of the `if` statement, neither version is supplied by the user. So we are comparing things like `foo@abcdef...` and `foo@abcdef...`, they are equal if the git IDs are equal as presented by the user.

There *is* a bug if the lookups would render to the same git ID: (e.g. a commit and a tag) e.g. `foo@abcdef...` and `foo@git.tag1` (if `tag1` resolves to `abcdef...`), although that's separate from what you mention here.